log-manager/src/main/java/io/airlift/log/BufferedHandler.java

```diff
@@ -368,13 +416,39 @@ public void flush()
         flushMessageOutputSafe();
     }
 
+    /**
+     * Requests a full flush of all currently queued messages from the background thread. If the handler has already been closed,
+     * The returned {@link ListenableFuture} will already be completed as no more background log processing can occur. Otherwise,
+     * the returned future will complete when the background flush has completed.
+     */
+    public ListenableFuture<Void> requestFullFlush()
+    {
+        SettableFuture<Void> flushedSignal;
+        queueDrainLock.lock();
+        try {
+            if (inputClosed.get() || terminalMessageDequeued) {
+                flushedSignal = requireNonNull(terminalMessage.flushSignal(), "terminalMessage flush signal must not be null");
+            }
+            else {
+                flushedSignal = SettableFuture.create();
+                queueInsert(new MessageAndSource(EMPTY_BYTES, "", flushedSignal));
```

Comment:
This calls `queueInsert` while holding the lock.  That method does a notify outside of the lock, so I think we need to do the insert outside of the drain lock.